File: Meta/canFinish.js
function canFinish(numCourses, prerequisites) {
  const graph = new Map();
  const visited = new Array(numCourses).fill(0); // State of each course

  // Build the graph
  for (const [course, prereq] of prerequisites) {
    if (!graph.has(prereq)) {
      graph.set(prereq, []);
    }
    graph.get(prereq).push(course);
  }

  // DFS to detect cycles
  const dfs = (course) => {
    if (visited[course] === 1) return false; // Cycle detected
    if (visited[course] === 2) return true; // Already processed

    visited[course] = 1; // Mark as visiting
    const neighbors = graph.get(course) || [];

    for (const neighbor of neighbors) {
      if (!dfs(neighbor)) {
        return false; // Cycle found
      }
    }

    visited[course] = 2; // Mark as visited
    return true;
  };

  // Check each course
  for (let i = 0; i < numCourses; i++) {
    if (!dfs(i)) {
      return false; // Cycle detected
    }
  }

  return true; // No cycles found, can finish all courses
}

// Time Complexity: O(V + E), where V is the number of courses and E is the number of prerequisites.
// Space Complexity: O(V + E), where V is the number of courses and E is the number of prerequisites.

File: Grind75/canFinish.py
def canFinish(numCourses, prerequisites):
    graph = [[] for _ in range(numCourses)]
    for dest, src in prerequisites:
        graph[src].append(dest)
    
    visited = [0] * numCourses # 0 = unvisited, 1 = visiting, 2 = visited

    def dfs(course):
        if visited[course] == 1:
            return False
        if visited[course] == 2:
            return True
        
        visited[course] = 1
        for neighbor in graph[course]:
            if not dfs(neighbor):
                return False
        visited[course] = 2
        return True
    
    for i in range(numCourses):
        if not dfs(i):
            return False
    
    return True